src/accountsdb/snapshots.zig

```diff
@@ -57,22 +110,73 @@ pub const Stakes = struct {
 
     /// history of staking levels
     stake_history: StakeHistory,
+
+    pub fn deinit(stakes: Stakes, allocator: std.mem.Allocator) void {
+        stakes.vote_accounts.deinit(allocator);
+
+        var stake_delegations = stakes.stake_delegations;
+        stake_delegations.deinit(allocator);
+
+        var stake_history = stakes.stake_history;
+        stake_history.deinit(allocator);
+    }
+
+    pub fn random(
+        allocator: std.mem.Allocator,
+        /// Should be a PRNG, not a true RNG. See the documentation on `std.Random.uintLessThan`
+        /// for commentary on the runtime of this function.
+        rand: std.Random,
+        max_list_entries: usize,
+    ) std.mem.Allocator.Error!Stakes {
+        const vote_accounts = try VoteAccounts.random(rand, allocator, max_list_entries);
+        errdefer vote_accounts.deinit(allocator);
+
+        var stake_delegations = try stakeDelegationsRandom(rand, allocator, max_list_entries);
+        errdefer stake_delegations.deinit(allocator);
+
+        var stake_history = try stakeHistoryRandom(rand, allocator, max_list_entries);
+        errdefer stake_history.deinit(allocator);
+
+        return .{
+            .vote_accounts = vote_accounts,
+
+            .stake_delegations = stake_delegations,
+            .unused = rand.int(u64),
+            .epoch = rand.int(Epoch),
+            .stake_history = stake_history,
+        };
+    }
 };
 
 /// Analogous to [VoteAccounts](https://github.com/anza-xyz/agave/blob/cadba689cb44db93e9c625770cafd2fc0ae89e33/vote/src/vote_account.rs#L44)
 pub const VoteAccounts = struct {
-    vote_accounts: std.AutoArrayHashMap(Pubkey, struct { u64, VoteAccount }),
-
-    staked_nodes: ?std.AutoArrayHashMap(
+    vote_accounts: std.AutoArrayHashMapUnmanaged(Pubkey, StakeAndVoteAccount),
+    staked_nodes: ?std.AutoArrayHashMapUnmanaged(
         Pubkey, // VoteAccount.vote_state.node_pubkey.
         u64, // Total stake across all vote-accounts.
     ) = null,
 
-    pub const @"!bincode-config:staked_nodes" = bincode.FieldConfig(?std.AutoArrayHashMap(Pubkey, u64)){ .skip = true };
+    pub const @"!bincode-config:staked_nodes" = bincode.FieldConfig(?std.AutoArrayHashMapUnmanaged(Pubkey, u64)){ .skip = true };
 
     const Self = @This();
 
-    pub fn stakedNodes(self: *Self, allocator: std.mem.Allocator) !*const std.AutoArrayHashMap(Pubkey, u64) {
+    pub const StakeAndVoteAccount = struct { u64, VoteAccount };
+
+    pub fn deinit(vote_accounts: VoteAccounts, allocator: std.mem.Allocator) void {
+        var copy = vote_accounts;
```

Comment:
why not just pass a mut reference?

Reply:
This is sort of an opportunistic preparatory step for @yewman's potential "deinit interface" - working with const deinits works a lot nicer with such things, since it means the code making use of that interface doesn't need to assume anything about the copyability of the values it's deinitializing.
Also just makes for nicer code where we can more liberally assign these to `const`s instead of `var`s that are never mutated, which also spread virally up through deinitialization methods - this is similar to the problem where a struct contains an RwMux field, and despite only reading from it, ends up needing to take a mut ptr anyway to interact with it.

Reply:
huh interesting i think i missed the deinit interface idea - will resolve for now so merge isnt blocked